File: clojure/word-count/src/word_count.clj
(ns word-count
  (:require [clojure.string :as str]))

(defn filter-characters [chars]
  (apply str
         (filter #(or (Character/isSpace %)
                      (Character/isDigit %)
                      (Character/isLetter %)) chars)))

(defn prepare-string [string]
  (-> string
      filter-characters
      str/lower-case
      (str/split #" +")))

(defn count-word [counter word]
  (let [current (get counter word)]
    (if (nil? current)
      (assoc counter word 1)
      (assoc counter word (inc current)))))

(defn word-count [words]
  (reduce count-word {} (prepare-string words)))

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map
  def count(sentence) do
    sentence
    |> String.downcase
    |> String.split(~r/\s|_|\W\W/, trim: true)
    |> Enum.reduce(%{}, fn(word, counter) -> Map.update(counter, word, 1, &(&1 + 1))
    end)
  end
end
